File: apps/macos/Sources/InvokeRenderer/ViewModel.swift
import Foundation
import InvokeIPC

/// The native side of the render-mutation stream (PLAN.md §4.5/§4.7). Applies the
/// same mutations the Node demo applies, but here these view models bind to AppKit
/// widgets (InvokePalette). Decoupling view models from concrete widgets lets the
/// wire format evolve without rewriting view code.
public final class ViewNode {
    public let id: Int
    public var type: String
    public var text: String?
    public var props: [String: JSONValue]
    public var children: [ViewNode] = []

    public init(id: Int, type: String, text: String? = nil, props: [String: JSONValue] = [:]) {
        self.id = id
        self.type = type
        self.text = text
        self.props = props
    }

    public var title: String? { props["title"]?.stringValue ?? text }
}

public final class ViewTree {
    public let root = ViewNode(id: 0, type: "#root")
    private var index: [Int: ViewNode] = [:]

    public init() { index[0] = root }

    public func apply(_ ops: [Mutation]) {
        for op in ops {
            switch op.op {
            case "clearContainer":
                root.children = []
            case "createInstance":
                if let id = op.id, let type = op.type {
                    index[id] = ViewNode(id: id, type: type, props: op.props ?? [:])
                }
            case "createText":
                if let id = op.id {
                    index[id] = ViewNode(id: id, type: "", text: op.text)
                }
            case "appendChild":
                if let p = index[op.parent ?? -1], let c = index[op.child ?? -1] { p.children.append(c) }
            case "insertBefore":
                if let p = index[op.parent ?? -1], let c = index[op.child ?? -1] {
                    if let b = index[op.before ?? -1], let at = p.children.firstIndex(where: { $0 === b }) {
                        p.children.insert(c, at: at)
                    } else {
                        p.children.append(c)
                    }
                }
            case "removeChild":
                if let p = index[op.parent ?? -1] { p.children.removeAll { $0 === index[op.child ?? -1] } }
            case "updateProps":
                if let id = op.id, let n = index[id] { n.props = op.props ?? [:] }
            case "updateText":
                if let id = op.id, let n = index[id] { n.text = op.text }
            default:
                break
            }
        }
    }

    public func describe() -> String {
        func walk(_ n: ViewNode, _ depth: Int) -> String {
            let pad = String(repeating: "  ", count: depth)
            let label = n.type.isEmpty ? "\"\(n.text ?? "")\"" : "\(n.type)\(n.title.map { " \"\($0)\"" } ?? "")"
            return ([pad + label] + n.children.map { walk($0, depth + 1) }).joined(separator: "\n")
        }
        return walk(root, 0)
    }
}

File: runtime/node-host/src/viewmodel.ts
/**
 * View-model tree — the host's side of the render-mutation stream (PLAN.md §4.5/§4.7).
 *
 * In production this lives in Swift and binds to AppKit widgets. Here it's a plain
 * tree so we can prove the protocol end-to-end in Node and pretty-print the result.
 */
import type { Mutation, NodeId, SerializedProps } from "@invoke/schema";
import { ROOT } from "@invoke/schema";

export interface VNode {
  id: NodeId;
  type: string; // "" for text nodes, "#root" for the container
  text?: string;
  props: SerializedProps;
  children: VNode[];
}

export interface ViewTree {
  root: VNode;
  index: Map<NodeId, VNode>;
}

export function createViewTree(): ViewTree {
  const root: VNode = { id: ROOT, type: "#root", props: {}, children: [] };
  return { root, index: new Map([[ROOT, root]]) };
}

export function applyMutations(tree: ViewTree, ops: Mutation[]): void {
  for (const op of ops) {
    switch (op.op) {
      case "clearContainer":
        // Only detach the container's children — instances created this commit
        // still live in the index and are about to be (re)attached.
        tree.root.children = [];
        break;
      case "createInstance":
        tree.index.set(op.id, { id: op.id, type: op.type, props: op.props, children: [] });
        break;
      case "createText":
        tree.index.set(op.id, { id: op.id, type: "", text: op.text, props: {}, children: [] });
        break;
      case "appendChild": {
        const parent = tree.index.get(op.parent);
        const child = tree.index.get(op.child);
        if (parent && child) parent.children.push(child);
        break;
      }
      case "insertBefore": {
        const parent = tree.index.get(op.parent);
        const child = tree.index.get(op.child);
        const before = tree.index.get(op.before);
        if (parent && child) {
          const at = before ? parent.children.indexOf(before) : -1;
          if (at >= 0) parent.children.splice(at, 0, child);
          else parent.children.push(child);
        }
        break;
      }
      case "removeChild": {
        const parent = tree.index.get(op.parent);
        const child = tree.index.get(op.child);
        if (parent && child) parent.children = parent.children.filter((c) => c !== child);
        break;
      }
      case "updateProps": {
        const n = tree.index.get(op.id);
        if (n) n.props = op.props;
        break;
      }
      case "updateText": {
        const n = tree.index.get(op.id);
        if (n) n.text = op.text;
        break;
      }
    }
  }
}

/** Find the first node of `type` whose `props.title` matches (used by the demo). */
export function findNode(tree: ViewTree, type: string, title?: string): VNode | undefined {
  for (const n of tree.index.values()) {
    if (n.type === type && (title === undefined || n.props.title === title)) return n;
  }
  return undefined;
}

const LABEL_PROPS = ["title", "subtitle", "text", "placeholder", "searchBarPlaceholder", "markdown"];

export function renderTree(node: VNode, depth = 0): string {
  const pad = "  ".repeat(depth);
  let line: string;
  if (node.type === "#root") {
    line = `${pad}◆ surface`;
  } else if (node.type === "") {
    line = `${pad}"${node.text}"`;
  } else {
    const label = LABEL_PROPS.map((p) => node.props[p]).find((v) => typeof v === "string");
    const extra: string[] = [];
    if (Array.isArray(node.props.accessories)) {
      const tags = (node.props.accessories as Array<Record<string, unknown>>)
        .map((a) => a.tag ?? a.text)
        .filter(Boolean);
      if (tags.length) extra.push(`[${tags.join(", ")}]`);
    }
    if (node.props.onAction) extra.push("⏎");
    line = `${pad}${node.type}${label ? ` "${String(label).slice(0, 40)}"` : ""}${extra.length ? " " + extra.join(" ") : ""}`;
  }
  const childLines = node.children.map((c) => renderTree(c, depth + 1));
  return [line, ...childLines].join("\n");
}
